### ISSUE TITLE
S3 error: Unable to verify integrity of data upload

### ISSUE THREAD

**Issue by JeromeSimmonds** [localstack/localstack#359]
Hi,

I am getting this error while trying to upload a file to the localstack S3:
`com.amazonaws.SdkClientException: Unable to verify integrity of data upload.
  Client calculated content hash (contentMD5: x5iB1Sj/A9YrpEmZCw2RpA== in base 64) didn't match hash (etag: 67393e672a1edc447ff78cf76d872684 in hex) calculated by Amazon S3.  You may need to delete the data stored in Amazon S3.`

Here is my code, which works with AWS:
```
...
String dataString = "{}"; // Some JSON content, just an example
byte[] dataBytes = dataString.getBytes(StandardCharsets.UTF_8);

ObjectMetadata metaData = new ObjectMetadata();
metaData.setContentType(ContentType.APPLICATION_JSON.toString());
metaData.setContentEncoding(StandardCharsets.UTF_8.name());
metaData.setContentLength(dataBytes.length);

byte[] resultByte = DigestUtils.md5(dataBytes);
String streamMD5 = new String(Base64.encodeBase64(resultByte));
metaData.setContentMD5(streamMD5);

PutObjectRequest putObjectRequest = new PutObjectRequest(bucketName, keyName, new ByteArrayInputStream(dataBytes), metaData);
s3Client.putObject(putObjectRequest);
 ```
I also tried without setting `metaData.setContentMD5(streamMD5);`, same error.

Thanks!

EDIT:
I forgot to say I use the Docker image localstack/localstack:latest and my Java app uses aws-java-sdk 1.11.104 (I also tried 1.11.80).

**Comment by whummer** [comment]
Thanks for reporting @JeromeSimmonds . I've been unable to reproduce this issue. Your test has been added to this class: https://github.com/localstack/localstack/blob/master/localstack/ext/java/src/test/java/cloud/localstack/S3UploadTest.java

```
make test-java
```

How exactly are you starting the Docker container - are you using any custom configuration parameters?

Would be great if you can extend the above test case to make the test *fail*, then we can further investigate this issue. Thanks

**Comment by JeromeSimmonds** [comment]
@whummer here is a snippet from my docker-compose file:
``` 
aws:
    image: localstack/localstack:latest
    ports:
      - "4572:4572"
```
Also here are more details about how I create the S3 client:
```
AWSCredentialsProvider awsCredentialsProvider = new AWSStaticCredentialsProvider(new BasicAWSCredentials(awsAccessKey, awsSecretKey));
AmazonS3ClientBuilder s3ClientBuider = AmazonS3ClientBuilder.standard()
    .withCredentials(awsCredentialsProvider);
...
s3ClientBuider.setEndpointConfiguration(new EndpointConfiguration(awsS3Endpoint, awsRegion));
s3ClientBuider.setPathStyleAccessEnabled(true);
s3Client = s3ClientBuider.build();
```
Thanks!

**Comment by jhovell** [comment]
@whummer @JeromeSimmonds  I may be seeing a different issue but in either case it seems there is an issue that uploads over >128k do not work with the same error as being reported here. I get a message such as:
```
Unable to verify integrity of data upload.  Client calculated content hash (contentMD5: 8eET31STItFmGOd/IrNlCA== in base 64) didn't match hash (etag: 7eebd4a71c4c983bd38c1a227c22b236 in hex) calculated by Amazon S3.  You may need to delete the data stored in Amazon S3. (metadata.contentMD5: 8eET31STItFmGOd/IrNlCA==, md5DigestStream: null, bucketName: 4d52e9a7-8666-4149-9ff8-70d22e035782, key: 187c0368-1f2b-4062-883a-cf40ccfff568)
```
Additionally if I suppress the exception and continue anyway the assert on the lengths of the returned object differs... so it would appear uploads over 128k are indeed getting corrupted:
``` 
java.lang.AssertionError: array lengths differed, expected.length=156482 actual.length=156478
	at org.junit.Assert.fail(Assert.java:88)
       ...
```

I can modify the test case (or maybe better yet add a new test case) so you can see what I am talking about.

This does seem to be a Java SDK specific issue. Or at least I have no problem uploading >128k objects using the AWS CLI.

Have tested with both Java 8 and Java 9. aws Java SDK version 1.11.215 (latest on Maven Central as of today).

**Comment by JeromeSimmonds** [comment]
Related issue https://github.com/localstack/localstack/issues/98
Seems like it's fixed now (at least for my use case).

**Comment by whummer** [comment]
Thanks for confirming @JeromeSimmonds . Closing this issue now, please re-open if you still experience this problem @jhovell

**Comment by aecc** [comment]
I'm having a similar issue using the JUnitRunner:
`Unable to verify integrity of data upload.  Client calculated content hash (contentMD5: 1B2M2Y8AsgTpgAmY7PhCfg== in base 64) didn't match hash (etag: 3c3e6a2d657082c14a37e8003898fb68 in hex) calculated by Amazon S3.  You may need to delete the data stored in Amazon S3. (metadata.contentMD5: null, md5DigestStream: com.amazonaws.services.s3.internal.MD5DigestCalculatingInputStream@3f5b2857, bucketName: test, key: tmp/item/8f314e45-f1ed-4012-ab7d-620e4c27763b/_temporary/0/)
com.amazonaws.SdkClientException: Unable to verify integrity of data upload.  Client calculated content hash (contentMD5: 1B2M2Y8AsgTpgAmY7PhCfg== in base 64) didn't match hash (etag: 3c3e6a2d657082c14a37e8003898fb68 in hex) calculated by Amazon S3.  You may need to delete the data stored in Amazon S3. (metadata.contentMD5: null, md5DigestStream: com.amazonaws.services.s3.internal.MD5DigestCalculatingInputStream@3f5b2857, bucketName: test, key: tmp/item/8f314e45-f1ed-4012-ab7d-620e4c27763b/_temporary/0/)`

Is this patch applied to the latest maven localstack-utils as well?

**Comment by aecc** [comment]
I'm trying to create a folder in S3, it would be an object of empty content: 
`
final ObjectMetadata om = new ObjectMetadata();
    om.setContentLength(0L);
`
Is this case handled in the localstack S3 service?

**Comment by nadim-haque-giffgaff** [comment]
Hi,

Just wanted to know if there was an update to this issue? Thanks.

**Comment by matthewmorales** [comment]
seems similar to https://github.com/jubos/fake-s3/issues/164 as the suggestion [here](https://github.com/jubos/fake-s3/issues/164#issuecomment-257305289) worked for me.

**Comment by blakeney** [comment]
For anyone reaching this issue thread via Internet search or otherwise still having similar problems, here's a suggestion:

Make sure that the file/data that you're uploading does not have a trailing newline.  I encountered an MD5 mismatch problem when using the S3 service in image version 0.9.4 of localstack, and I worked around it by removing a newline that my editor had automatically placed at the end of the file containing the test data I was trying to upload.

I do not know whether this is an actual issue in localstack, or whether it's specific to my environment, but I'm sharing the information here in case it helps someone.
